Sources/AWSLambdaRuntimeCore/Documentation.docc/Proposals/0001-v2-api.md

```diff
@@ -0,0 +1,556 @@
+
+## Detailed Solution
+
+Below you’ll find explanation’s for all types that we want to use in AWS Lambda Runtime v2.
+
+### LambdaResponseWriter
+
+We will introduce a new `LambdaResponseWriter` protocol. It is used in the new `StreamingLambdaHandler` (defined below), which is the new base protocol for the `LambdaRuntime` (defined below as well).
+
+```swift
+/// A writer object to write the Lambda response stream into
+public protocol LambdaResponseWriter: ~Copyable {
+    /// Write a response part into the stream
+    func write(_ buffer: ByteBuffer) async throws
+    /// End the response stream
+    consuming func finish()
+    /// Write a response part into the stream and end the response stream
+    consuming func writeAndFinish(_ buffer: ByteBuffer) async throws
+    /// Report an error in the response stream
+    consuming func reportError(_ error: any Error) async throws
+}
+```
+
+### LambdaContext
+
+`LambdaContext` will be largely unchanged, but the `eventLoop` property as well as the `allocator` property (of type `ByteBuffer`) will be removed.
+
+A new function `backgroundTask()` will also be added. This will allow tasks to be run in the background while and after the response is/has been sent. Please note that `LambdaContext` will not be Sendable anymore.
+
+```swift
+/// A context object passed as part of an invocation in LambdaHandler handle functions.
+public struct LambdaContext {
+    /// The request ID, which identifies the request that triggered the function invocation.
+    public var requestID: String { get }
+
+    /// The AWS X-Ray tracing header.
+    public var traceID: String { get }
+
+    /// The ARN of the Lambda function, version, or alias that's specified in the invocation.
+    public var invokedFunctionARN: String { get }
+
+    /// The timestamp that the function times out.
+    public var deadline: DispatchWallTime { get }
+
+    /// For invocations from the AWS Mobile SDK, data about the Amazon Cognito identity provider.
+    public var cognitoIdentity: String? { get }
+
+    /// For invocations from the AWS Mobile SDK, data about the client application and device.
+    public var clientContext: String? { get }
+
+    /// `Logger` to log with.
+    ///
+    /// - note: The `LogLevel` can be configured using the `LOG_LEVEL` environment variable.
+    public var logger: Logger { get }
+
+    /// Schedule background work in lambda. This work must be completed before
+    /// the runtime can ask for further events. Note that AWS will continue to
+    /// charge you after the response has been returned, but work is still
+    /// processing in the background.
+    public func addBackgroundTask(_ body: sending @escaping () async -> ())
```

Comment:
This doesn't feel necessary to me with the current API proposal. Since we pass in an independent writer. A user can just write some bytes, the call writer then finish and then do any background work before returning from the `handle` method. They can even set that up in child tasks in the `handle` method to start the background work while handling the actual request.

Reply:
Agree. This is unstructured concurrency and given that the user controls `finish`, they know when the response is done.

Reply:
@weissi this would not be unstructured, as the underlying runtime would inject a `taskGroup` into the context and that `taskGroup` would be used to schedule the child task here. The underlying runtime would then ensure that all tasks have completed, before it would ask for more work. This approach would be necessary, if we used an approach in which the user returns a LambdaResponse:

```swift
protocol LambdaStreamingResponse {
    func handle(_ event: ByteBuffer, context: LambdaContext) async throws -> LambdaResponse
}
```

However this _is_ needed for the `LambdaHandler` API:

```swift
public protocol CodableLambdaHandler: ~Copyable {
    associatedtype Event
    associatedtype Output

    // only way to schedule background work that can continue after returning the Output here
    // is by having the addBackgroundTask on `LambdaContext` here.
    func handle(_ event: Event, context: LambdaContext) async throws -> Output
}
```

@FranzBusch given this, do we want to have two different `LambdaContext`s? Or are we fine with keeping the `LambdaContext.addBackgroundTask` even in the situation where it isn't really needed? I think I would opt for keeping it (so that we have it for the 99% use-case)

Reply:
@fabianfett I'm pretty sure that it's is unstructured, it just manages to now use a `Task {` to achieve it. It passes the background work into a reference which will then be owned by something _further up the stack_.

The test is easy, if you write this code

```
do {
    thing.addBackgroundWork { print("AAAA") }
} while false
print("BBBB")
```

Is there any way in which `AAAA` could print _after_ `BBBB`? If so: unstructured.

Reply:
Structured Concurrency means that when you reach the next `}`, then all the work you have scheduled must have terminated or wait for it. There's no going further until that's complete.

Reply:
@weissi Okay I get your point. 

But the point of adding this API is that users explicitly ask for it #334 and AWS makes a deal about supporting background work: [Running code after returning a response from an AWS Lambda function](https://aws.amazon.com/blogs/compute/running-code-after-returning-a-response-from-an-aws-lambda-function/)

In Lambdas case this can be used to flush metrics or logs _after_ the response has returned. So from a structured concurrency standpoint, yes this is flawed! But it is a useful feature nonetheless. 

We do use structured concurrency in the level above to ensure all background tasks finish before we get the next invocation.

Reply:
Can't we express this like so:

```
func handle() {
    try await writeResponse()
    try await flushMetrics()
}
```

That way we can run code after writing a response and we're also structured. You could also provide a library function that performs certain jobs after the response has been written.

Reply:
@weissi agree that this works for the Streaming case. However I think we should keep the easy Codable support API:

```
public protocol LambdaHandler: ~Copyable {
    associatedtype Event
    associatedtype Output

    // only way to schedule background work that can continue after returning the Output here
    // is by having the addBackgroundTask on `LambdaContext` here.
    func handle(_ event: Event, context: LambdaContext) async throws -> Output
}
```

In that case you can't `flushMetrics()` after the response was returned. And the `addBackgroundTask()` would be handy here.

Reply:
why don't you make it

```
public protocol LambdaHandler: ~Copyable {
    associatedtype Event
    associatedtype Output

    func handle(_ event: Event, context: LambdaContext) async throws -> Output

    func performPostHandleWork(...) async throws -> Void
} 

extension LambdaHandler {
    func performPostHandleWork(...) async throws -> Void {
        // nothing to do
    }
}
```

that way the users can implement post-handle work

Reply:
The problem with splitting it across two methods is if the post work needs context from the initial work. However, you could also just pass a writer that takes the output to the `handle` method to allow for background work. It's not as nice as returning the Output from the method.

Reply:
> The problem with splitting it across two methods is if the post work needs context from the initial work. 

but we can pass the context, no?

Reply:
@fabianfett Not sure I understand what "a task that starts immediately but is allowed to run longer" is. Could you provide some example code?

Reply:
An API call that 

1. does a short computation which is returned to you immediately 
2. spawns a longer computation, which's result is send to you via mail once it finishes (can take up to 15min)

Both should start as soon as the API is invoked. We can potentially question the API design here. But this is possible with Lambda and we should allow users to do this, if they opt to.

Reply:
Probably the most consistent approach for @fabianfett's use case is to provide a wrapping `CodableResponseWriter` (potentially with just a `writeAndFinish` API) while allowing work to continue past the result being returned to the control plane. It is not as simple as returning the result directly. We **could** provide both forms but that does add its own complexity to the library overall.